functorch/experimental/_scan.py

```diff
@@ -0,0 +1,370 @@
+from functools import partial
+import torch
+from torch._subclasses.fake_tensor import FakeTensorMode
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._functorch.aot_autograd import create_joint, AOTConfig
+from torch._ops import HigherOrderOperator
+from torch.multiprocessing.reductions import StorageWeakRef
+from torch.fx.experimental.proxy_tensor import (
+    ProxyTorchDispatchMode,
+    disable_proxy_modes_tracing,
+    make_fx,
+    track_tensor_tree,
+    unwrap_proxy,
+)
+from torch.utils._python_dispatch import (
+    _get_current_dispatch_mode,
+    _pop_mode_temporarily,
+)
+from torch._dispatch.python import suspend_functionalization
+
+
+# TODO: We add this to prevent dymamo from tracing into map_wrapper,
+# remove the wrapper call when it's ready.
+class ScanWrapper(HigherOrderOperator):
+    def __call__(self, f, init, xs):
+        return scan_wrapper(f, init, xs)
+
+scan = ScanWrapper("scan")
+scan_impl = HigherOrderOperator("scan_impl")
+
+dummy_aot_config = AOTConfig(fw_compiler=None,
+                             bw_compiler=None,
+                             partition_fn=None,
+                             decompositions={},
+                             num_params_buffers=0,
+                             aot_id=0,
+                             keep_inference_input_mutations=False)
+
+
+def create_fw_bw_graph(f, flat_init, flat_xs):
+    # Note: We create "clean" environments for make_fx by suspending all dispatch keys
+    # between Autograd and Python key. Currently, we only suspend functionalization but more can be
+    # added when required. Will encounter two problems if we don't suspend functionalization:
+    #
+    # 1. make_fx fails to capture operations on input: the inputs are wrapped as _to_functional_tensor_wrapper,
+    # but they will be unwrapped before entering ProxyTorchDispatchMode as part of the dispatching.
+    # However, it's the outside wrapper that tracer creates proxies for. This casuses tracer fail to
+    # fetch the proxy for the inputs and fail to capture any operations on them.
+    #
+    # 2. make_fx fails to capture output: the outputs after ProxyTorchDispatchMode are further
+    # wrapped as FunctionalTensorWrapper in Functionalize key after return. However, the tracer
+    # only associates the inner tensor with proxy in ProxyTorchDispatchMode. Therefore,
+    # when creating the output node, it fails to associate the wrapped tensor with its proxy.
+    # Instead, it will create _tensor_constant as output.
+
+    with suspend_functionalization():
+        with disable_proxy_modes_tracing():
+            def from_fun(t):
+                if isinstance(t, torch.Tensor):
```

Comment:
Is this function applicable for map/cond/etc as well? If so we should try to share code between the three implementations.

It sounds like we just want a special version of `make_fx` that:
1. suspends dispatcher state
2. also gets rid of all Tensor subclasses in the input